File: src/game/whackamole.ts
import { Neovim } from "neovim";
import { GameBuffer } from "../game-buffer";
import { GameState, GameOptions } from "./types";
import { BaseGame, getRandomSentence } from "./base";

export class WhackAMoleGame extends BaseGame {
    private instructionLines = [
        "How to Play:",
        "------------",
        "",
        "Use vim movements to locate the character with the arrow under it as ",
        "quickly as possible. Then invert the character's case to win."
    ];
    private winLine: string;
    private outputStartRow = 2;

    constructor(nvim: Neovim, buffer: GameBuffer, state: GameState, opts?: GameOptions) {
        super(nvim, buffer, state, opts);

        this.winLine = "";
    }

    // I think I could make this all abstract...
    async hasFailed(): Promise<boolean> {
        return false;
    }

    async run(): Promise<void> {
        const sentence = getRandomSentence();
        let chosenLocation: number = 0;

        do {
            const location = Math.floor(Math.random() * sentence.length);

            if (location > 0 && /[A-Za-z]/.test(sentence[location])) {
                chosenLocation = location;
            }
        } while (!chosenLocation);

        const pointerLine = sentence
            .split("")
            .map((_, index) => (index === chosenLocation ? "^" : " "))
            .join("");

        this.winLine = this.createWinLine(sentence, chosenLocation);

        await this.state.buffer.setLines(
            [...this.instructionLines, "", sentence, pointerLine],
            {
                start: this.outputStartRow,
                strictIndexing: true
            }
        );

        this.state.window.cursor = [
            this.instructionLines.length + this.outputStartRow + 1,
            0
        ];
    }

    async clear(): Promise<void> {
        const len = await this.state.buffer.length;
        await this.state.buffer.remove(0, len, true);
        await this.state.buffer.insert(
            new Array(this.state.lineRange.end).fill(""),
            0
        );
    }

    async gameOver(): Promise<void> {
        await this.clear();
    }

    async checkForWin(): Promise<boolean> {
        const lines = await this.state.buffer.getLines({
            start: this.instructionLines.length + this.outputStartRow + 1,
            end: this.instructionLines.length + this.outputStartRow + 2,
            strictIndexing: false
        });

        return lines[0] === this.winLine;
    }

    private createWinLine(sentence: string, targetColumn: number): string {
        const preTargetChar = sentence.slice(0, targetColumn);
        const postTargetChar = sentence.slice(targetColumn + 1);
        let targetChar = sentence[targetColumn];

        if (/[a-z]/.test(targetChar)) {
            targetChar = targetChar.toUpperCase();
        } else {
            targetChar = targetChar.toLowerCase();
        }

        return preTargetChar + targetChar + postTargetChar;
    }
}

File: lua/vim-be-good/games/whackamole.lua
local GameUtils = require("vim-be-good.game-utils")
local log = require("vim-be-good.log")

local instructions = {
    "Use vim movements to locate the character with the arrow under it as ",
    "quickly as possible. Then invert the character's case to win.",
    "",
}

local WhackAMoleRound = {}
function WhackAMoleRound:new(difficulty, window)
    log.info("New", difficulty, window)
    local round = {
        window = window,
        difficulty = difficulty,
    }

    self.__index = self
    self.__winLine = ""
    return setmetatable(round, self)
end

function WhackAMoleRound:getInstructions()
    return instructions
end

function WhackAMoleRound:getConfig()
    log.info("getConfig", self.difficulty, GameUtils.difficultyToTime[self.difficulty])
    return {
        roundTime = GameUtils.difficultyToTime[self.difficulty]
    }
end

function WhackAMoleRound:checkForWin()
    local lines = self.window.buffer:getGameLines()
    local found = false
    found = lines[1] == self.__winLine
    return found
end

function WhackAMoleRound:render()
    local sentence = GameUtils:getRandomSentence()
    local lines = GameUtils.createEmpty(2)

    lines[1] = sentence

    local foundLocation = false
    local chosenLocation = 0
    while not foundLocation do
        local location = math.random(1, string.len(sentence))
        if sentence:sub(location,location) ~= " " then
            chosenLocation = location
            foundLocation = true
        end
    end

    local pointerLine = ""
    local winSentence = ""
    for idx=1, #sentence do
        local c = sentence:sub(idx,idx)
        if idx == chosenLocation then
            pointerLine = pointerLine .. "^"
            if c == string.lower(c) then
                winSentence = winSentence .. string.upper(c)
            else
                winSentence = winSentence .. string.lower(c)
            end
        else
            pointerLine = pointerLine .. " "
            winSentence = winSentence  .. c
        end
    end
    lines[2] = pointerLine

    self.__winLine = winSentence

    local cursorIdx =  1

    return lines, cursorIdx
end

function WhackAMoleRound:name()
    return "whackamole"
end

return WhackAMoleRound


